diff --git a/src/arch/generic/packedpair.rs b/src/arch/generic/packedpair.rs
index 8d97cf28..45d088cf 100644
--- a/src/arch/generic/packedpair.rs
+++ b/src/arch/generic/packedpair.rs
@@ -244,7 +244,7 @@ impl<V: Vector> Finder<V> {
         while offsets.has_non_zero() {
             let offset = offsets.first_offset();
             let cur = cur.add(offset);
-            if end.sub(needle.len()) < cur {
+            if end.distance(cur) < needle.len() {
                 return None;
             }
             if is_equal_raw(needle.as_ptr(), cur, needle.len()) {
diff --git a/src/arch/x86_64/sse2/packedpair.rs b/src/arch/x86_64/sse2/packedpair.rs
index c8b5b999..75329c2c 100644
--- a/src/arch/x86_64/sse2/packedpair.rs
+++ b/src/arch/x86_64/sse2/packedpair.rs
@@ -229,4 +229,25 @@ mod tests {
         }
         crate::tests::packedpair::Runner::new().fwd(find).run()
     }
+
+    // A demonstration of UB (caught under Miri) through some of memchr's
+    // lower level but public APIs.
+    //
+    // See: https://github.com/BurntSushi/memchr/issues/225
+    #[test]
+    fn regression_undefined_behavior() {
+        let construct_needle: &[u8] = b"ab"; // length 2 → pair indices 0,1
+        let pair = Pair::new(construct_needle).unwrap();
+        let Some(finder) = Finder::with_pair(construct_needle, pair) else {
+            return;
+        };
+        let mhl = finder.min_haystack_len(); // 17
+        let mut haystack = alloc::vec![b'.'; mhl + 5];
+        haystack[0] = b'a';
+        // force entry into find_in_chunk inner branch
+        haystack[1] = b'b';
+        // 122 > haystack.len()=22
+        let long_needle = alloc::vec![b'X'; haystack.len() + 100];
+        let _ = finder.find(&haystack, &long_needle);
+    }
 }
